src/timely-util/src/temporal.rs

```diff
@@ -0,0 +1,328 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License in the LICENSE file at the
+// root of this repository, or online at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Utilities to efficiently store future updates.
+
+use std::collections::BTreeMap;
+
+use mz_ore::cast::CastFrom;
+use timely::progress::Timestamp;
+use timely::progress::frontier::AntichainRef;
+
+/// Timestamp extension for timestamps that can advance by `2^exponent`.
+///
+/// Most likely, this is only relevant for totally ordered timestamps.
+pub trait BucketTimestamp: Timestamp {
+    /// The number of bits in the timestamp.
+    const DOMAIN: usize = size_of::<Self>() * 8;
+    /// Advance this timestamp by `2^exponent`.
+    fn advance_by_power_of_two(&self, exponent: u32) -> Option<Self>;
+}
+
+/// A type that can be split into two parts based on a timestamp.
+pub trait Storage: Sized {
```

Comment:
Nit: Consider calling this `Bucket`. I think that would make the usage of this trait more obvious, given that it's a part of a `BucketChain` and contains `BucketTimestamp`s.